docs/src/main/mdoc/datatypes/statet.md

```diff
@@ -0,0 +1,177 @@
+                                                    ^                                             |
+                                                    |                                             |
+                                                    |                                             |
+                  +-----------+          +---------------+       +-----------------------+        |
+                  | Game Over |<---------+ Check the win |<------| Compute the new state |<-------+
+                  +-----------+          |   condition   |       +-----------------------+
+                                         +---------------+
+
+```
+
+At each step, the state might be updated and, since some I/O operations needs to be performed along the way, the specific effect (typically `IO` monad, but in our case it will be `Eval`) has to be used.
+We can model the state as:
+
+```scala mdoc:silent
+import cats.data.StateT
+import cats.Eval
+import scala.io.StdIn._
+
+final case class GameState(
+    target: String,
+    guessedChars: Set[Char],
+    attemptNum: Int
+)
+
+sealed trait GameOver
+case object Win extends GameOver
+case object Loose extends GameOver
+
+object GameState {
+
+  val maxAttempts: Int = 3
+
+  def apply(target: String): GameState = GameState(
+    target = target,
+    guessedChars = Set.empty[Char],
+    attemptNum = 0
+  )
+
+  def showWordHidden(state: GameState): String =
+    state.target.map(c => if (state.guessedChars.contains(c)) c else '*')
+
+  def attemptGuess(
+      state: GameState,
+      guess: String
+  ): Either[GameOver, GameState] =
+    state match {
+      case GameState(t, _, _) if t == guess => Left(Win)
+      case GameState(_, _, attemptNum) if attemptNum == (maxAttempts - 1) =>
+        Left(Loose)
+      case GameState(target, guessedChars, attemptNum) =>
+        Right(
+          GameState(
+            target = target,
+            guessedChars = guessedChars ++ guess,
+            attemptNum = attemptNum + 1
+          )
+        )
+    }
+
+}
+```
+
+In the code above you can see some useful functions, such the one that returns the encrypted target word. A brief example here:
+
+```scala mdoc
+val gameState1 = GameState("cats")
+println(GameState.showWordHidden(gameState1))
+val gameState2 = GameState.attemptGuess(gameState1, "bats")
+println(GameState.showWordHidden(gameState2.getOrElse(???)))
+```
+
+Let's focus on the state machine game steps.
+
+```scala mdoc:silent
+object StateMachine {
+
+  def step1(): Eval[GameState] = for {
+    _ <- Eval.later(println("Insert the target word: "))
+    word <- Eval.later(readLine())
+  } yield GameState(word)
+
+  def step2(): StateT[Eval, GameState, Unit] = for {
+    state <- StateT.get[Eval, GameState]
+    _ <- StateT.liftF(Eval.later(println(GameState.showWordHidden(state))))
+  } yield ()
+
+  def step3(): StateT[Eval, GameState, String] = for {
+    _ <- StateT.liftF(Eval.later(println("Insert the guess word: ")))
```

Comment:
I also am -1 on using Eval as a poor man's IO monad.

We should not use a pattern we don't recommend in our documentation.

I would instead suggest using `StateT[Either[String, *], GameState, Unit]` in order to model steps that can somehow fail or something along those lines.

Reply:
`Either` seems good. Easier to grok than `Writer`.